simplesource-data/src/main/java/io/simplesource/data/CommandError.java

```diff
@@ -0,0 +1,158 @@
+     * @param ex the function that receives the reason and throwable, returning a value of the specified type
+     * @return the result
+     */
+    public abstract <A> A fold(BiFunction<Reason, String, A> str, BiFunction<Reason, Throwable, A> ex);
+
+    //
+    // internals
+    //
+
+    private final Reason reason;
+
+    private CommandError(final Reason reason) {
+        this.reason = reason;
+    }
+
+    @Override
+    public String toString() {
+        return "CommandError(" + fold(
+                (error, message) -> "Error: " + error + " Message: " + message,
+                (error, throwable) -> "Error: " + error + " Throwable: " + throwable) + ')';
+    }
+
+    static final class ThrowableCommandError extends CommandError {
+        private final Throwable throwable;
+
+        ThrowableCommandError(final Reason reason, final Throwable throwable) {
+            super(reason);
+            this.throwable = requireNonNull(throwable);
+        }
+
+        @Override
+        public String getMessage() {
+            return throwable.getMessage();
+        }
+
+        @Override
+        public <A> A fold(final BiFunction<Reason, String, A> str, final BiFunction<Reason, Throwable, A> ex) {
+            return ex.apply(getReason(), throwable);
+        }
+
+        @Override
+        public int hashCode() {
+            return 31 * getReason().hashCode() + throwable.hashCode();
+        }
+
+        @Override
+        public boolean equals(final Object obj) {
+            if (!(obj instanceof CommandError.ThrowableCommandError))
+                return false;
+            final ThrowableCommandError other = (ThrowableCommandError) obj;
+            return Objects.equals(getReason(), other.getReason()) &&
+                    Objects.equals(throwable, other.throwable);
+        }
+
+    }
+
+    static final class StringCommandError extends CommandError {
+        private final String msg;
+
+        StringCommandError(final Reason reason, final String msg) {
+            super(reason);
+            this.msg = requireNonNull(msg);
+        }
+
+        @Override
+        public String getMessage() {
+            return msg;
+        }
+
+        @Override
+        public <A> A fold(final BiFunction<Reason, String, A> str, final BiFunction<Reason, Throwable, A> ex) {
+            return str.apply(getReason(), msg);
+        }
+
+        @Override
+        public int hashCode() {
+            return 31 * getReason().hashCode() + msg.hashCode();
+        }
+
+        @Override
+        public boolean equals(final Object obj) {
+            if (!(obj instanceof CommandError.StringCommandError))
+                return false;
+            final StringCommandError other = (StringCommandError) obj;
+            return Objects.equals(getReason(), other.getReason()) &&
+                    Objects.equals(msg, other.msg);
+        }
+
+    }
+
+    public enum Reason {
```

Comment:
I'm thinking we should call this `ErrorCode` rather than `Reason`.

Reply:
I am ok with that, but then should we force a numeric or some kind of id on it?

Reply:
I think `CommandError.Reason` sounds better than `CommandError.ErrorCode`.